graph/generate.go

```diff
@@ -0,0 +1,254 @@
+/*
+Package graph provides core lineage graph services. It defines the graph datatype and provides the "New"
+function for creating one.
+*/
+package graph
+
+import (
+	"fmt"
+
+	mbp "github.com/GoogleCloudPlatform/healthcare-data-harmonization/mapping_engine/proto"
+	proto "github.com/golang/protobuf/proto"
+)
+
+var idFactory IDfactory = &autoIncFactory{currentID: 0}
+
+// setIDfactory sets the interface used for ID generation
+func setIDfactory(factory IDfactory) {
+	idFactory = factory
+}
+
+// msgContext is a wrapper around a proto message storing additional context, like parent node.
+type msgContext struct {
+	msg        proto.Message
+	descendant Node
+}
+
+type stack []msgContext
+
+func (s stack) push(mc msgContext) stack {
+	return append(s, mc)
+}
+
+// reversePush pushes an array of proto messages onto the stack in reverse order. It assigns them all the same context.
+// The order is reversed to preserve the ordering of locally-defined whistler targets. For example, an array
+// of FieldMappings extracted from a Projector
+func (s stack) reversePush(msgs []proto.Message, descendant Node) stack {
```

Comment:
nit: the APIs are better to be consistent, for example, `push` takes a msgContext, we should probably take a list of msgContext here.

Reply:
I renamed stack to indicate it is specific to building a graph using a DFS algorithm. I also made it always take an array of msgContext, and added a helper method to convert a list of whistler messages to msgContexts.